src/pokemon.rs

```diff
@@ -15,8 +19,176 @@ pub struct Pokemon {
     pub forms: Vec<String>,
 }
 
-pub fn load_pokemon(pokemon_db: &EmbeddedFile) -> Result<Vec<Pokemon>, Error> {
-    let pokemon_json_str = str::from_utf8(&pokemon_db.data).expect("Invalid UTF-8 in pokemon db");
-    let pokemon: Vec<Pokemon> = serde_json::from_str(pokemon_json_str)?;
-    Ok(pokemon)
+impl Pokemon {
+    /// Get all forms of the pokemon except those in the exclude vector
+    pub fn get_filtered_forms(&self, exclude: &[&str]) -> Vec<String> {
+        self.forms
+            .iter()
+            .filter(|f| !exclude.contains(&f.as_str()))
+            .cloned()
+            .collect()
+    }
+}
+
+trait GenerationParser {
+    fn parse_generations(&self) -> Result<(u8, u8), Error>;
+}
+
+impl GenerationParser for &str {
+    fn parse_generations(&self) -> Result<(u8, u8), Error> {
+        let (start_gen, end_gen) = match self.split_once('-') {
+            Some(gens) => gens,
+            None => {
+                let gen_list = self.split(',').collect::<Vec<_>>();
+                let gen = gen_list.choose(&mut rand::thread_rng()).unwrap();
+                (*gen, *gen)
+            }
+        };
+
+        let start_gen = start_gen.parse::<u8>();
+        let end_gen = end_gen.parse::<u8>();
+        match (start_gen, end_gen) {
+            (Ok(s), Ok(e)) => Ok((s, e)),
+            _ => Err(Error::InvalidGeneration(self.to_string())),
+        }
+    }
+}
+
+pub struct PokemonDatabase {
+    pokemon: Vec<Pokemon>,
+    config: Config,
+}
+
+impl PokemonDatabase {
+    pub fn load(pokemon_db: &EmbeddedFile, config: Config) -> Result<Self, Error> {
+        let pokemon_json_str =
+            str::from_utf8(&pokemon_db.data).expect("Invalid UTF-8 in pokemon db");
+        let pokemon: Vec<Pokemon> = serde_json::from_str(pokemon_json_str)?;
+        Ok(Self { pokemon, config })
+    }
+
+    /// Filter pokemon by generation
+    pub fn filter_by_generation(&self, generations: &str) -> Result<Vec<&Pokemon>, Error> {
+        let (start_gen, end_gen) = generations.parse_generations()?;
+
+        let pokemon: Vec<&Pokemon> = self
+            .pokemon
+            .iter()
+            .filter(|p| (start_gen..=end_gen).contains(&p.gen))
+            .collect();
+
+        if pokemon.is_empty() {
+            return Err(Error::InvalidGeneration(generations.to_string()));
+        }
+
+        Ok(pokemon)
+    }
+
+    /// Returns a vector of all pokemon
+    pub fn get_all(&self) -> &Vec<Pokemon> {
+        &self.pokemon
+    }
+
+    /// Prints the names of all pokemon from the given generations into stdout
+    pub fn list_pokemon_names(&self, generations: &str) -> Result<(), Error> {
+        let filtered_pokemon = self.filter_by_generation(generations)?;
```

Comment:
This does not seem to work with comma separated generations, eg krabby list 1,2 only returns a list of gen 1 pokemon

Reply:
Silly mistake 🤣 All good now.

Glad you're thorough!